### DISCUSSION TITLE
Jeedom Official Zigbee Plugin based on zigpy now listed as "stable"

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#725] [General]
FYI, just read the news that Jeedom Official Zigbee Plugin based on zigpy have now been listed as "stable" and as such is no longer in "beta":

https://blog.jeedom.com/5183-tout-ce-quil-faut-savoir-sur-le-plugin-officiel-zigbee/ <- **This announcement blog post is a great read!**

https://blog.jeedom.com/category/protocole/zigbee/

https://market.jeedom.com/index.php?v=d&p=market_display&id=4050

https://doc.jeedom.com/en_US/plugins/automation%20protocol/zigbee/

"_The Zigbee plugin for Jeedom builds on the excellent work done around the open-source Zigpy library to offer a general compatibility with different Zigbee hardware. It allows communication with the following Zigbee controllers_"

https://doc.jeedom.com/en_US/plugins/automation%20protocol/zigbee/changelog

https://community.jeedom.com/t/plugin-officiel-zigbee/58321

https://community.jeedom.com/t/plugin-zigbee-beta/40301

Translation: https://translate.google.com/translate?sl=fr&tl=en&u=https://community.jeedom.com/t/plugin-zigbee-beta/40301

https://community.jeedom.com/t/plugin-zigbee-beta-blabla/53420

https://community.jeedom.com/tag/plugin-zigbee

https://compatibility.jeedom.com/index.php?v=d&p=home&search=&lang=en_US&protocol=Zigbee

@zoic21 Any idea if and when the source code for that zigpy based "plugin-zigbee" for Jeedom will become available in a public repo?

https://github.com/jeedom

https://github.com/jeedom/zigpy

i.e. https://github.com/jeedom/plugin-zigbee now leads to "404 page not found message" while other plugins have public repos.

PS: I don't use Jeedom myself (and I don't speak French) but I see that the Jeedom core itself is based on JavaScript and PHP, not Python:

https://github.com/jeedom/core

https://www.jeedom.com/

**Comment by Hedda** [comment]
@zoic21 by the way, zigpy-znp is no longer considered as experimental and should from now be the prefered library for Z-Stack 3 and Z-Stack Home 1.2 adapters

https://github.com/home-assistant/home-assistant.io/pull/17089

https://github.com/zigpy/zigpy/pull/705

https://github.com/zigpy/zigpy-znp/pull/69

As such might want to update blog + documentation to no longer state that zigpy-znp is "in experimental in Zigpy" and remove mentioning of Z-Stack 3.x.x as zigpy-znp should be considered fully compatible with both Z-Stack 3.x.x and Z-Stack Home 1.2 here:

https://doc.jeedom.com/en_US/plugins/automation%20protocol/zigbee/

https://blog.jeedom.com/5183-tout-ce-quil-faut-savoir-sur-le-plugin-officiel-zigbee/

**Comment by Hedda** [comment]
@zoic21 Jeedom should really also promote the **zha-device-handlers** project specifically as well if your plugin is depending on it:

https://github.com/zigpy/zha-device-handlers -> https://github.com/zigpy/zha-device-handlers/blob/dev/README.md

Home Assistant's ZHA (Zigbee Home Automation) integration provides this information to point people to that project:

https://www.home-assistant.io/integrations/zha/#zha-exception-and-deviation-handling

### **_ZHA EXCEPTION AND DEVIATION HANDLING_**

_The ZHA implementation in Home Assistant relies on a library called “[ZHA Device Handlers](https://github.com/zigpy/zha-device-handlers)” to resolve issues with Zigbee devices that do not fully conform with the Zigbee standards. The few devices that deviate from the Zigbee specifications set by the [Zigbee Alliance](https://zigbeealliance.org/) may therefore require proper bug reports with debug logs from users to assistant the developers in writing custom [ZHA Device Handlers](https://github.com/zigpy/zha-device-handlers/blob/dev/README.md) for all of a device functions to work properly with the ZHA integration._

_Such a custom “ZHA Device Handler” are Python scripts that internally are also referred to as a “quirk” because they fix “quirks”, like deviations from the standard specifications. ZHA Device Handles do this by transparently, acting as a translator, translating and converting non-compliant device messages and instead present them to the application as coming from a virtual compliant device. These ZHA Device Handlers for Home Assistant can thus be used to parse custom messages to and from Zigbee devices. The ZHA Device Handlers that are made can then be reused by all users in future versions of Home Assistant._

_The custom quirks implementations for zigpy implemented as ZHA Device Handlers for Home Assistant are a similar concept to that of [Hub-connected Device Handlers for the SmartThings Classics platform](https://docs.smartthings.com/en/latest/device-type-developers-guide/) as well as that of [Zigbee-Herdsman Converters (formerly Zigbee-Shepherd Converters) as used by Zigbee2mqtt](https://www.zigbee2mqtt.io/how_tos/how_to_support_new_devices.html), meaning they are each virtual representations of a physical device that expose additional functionality that is not provided out-of-the-box by the existing integration between these platforms._

**Comment by Hedda** [comment]
@zoic21 Do you happen to have any contacts at Jeedom SAS and/or any other developers working on this Jeedom Zigbee Plugin?

Was just wondering if they maybe have any resources that could help write development upstream documentation for zigpy library?

Thinking that it would benefit Jeedom too if had more documentation about how to implement zigpy in an application like Jeedom.

I believe if more third-party applications that implement zigpy could contribute to it upstream then more projects will start using it.

One thing that could be holding that up is documentation on how to get started with zigpy. That is, a how-to begin starting guide.

See example: https://github.com/zigpy/zigpy/issues/715

**Comment by zoic21** [comment]
Hello,
Jeedom SAS does not already have enough time to make its own documents I doubt they will find any to do the Zigpy documentation...But it is sure that it would greatly benefit the zigpy project to have some documentation so I would ask them but I do not believe it

**Comment by Hedda** [comment]
FYI, noticed that Loic posted a new blog-post that explains different Zigbee solutions for Jeedom and why Zigbee is hard to implement:

https://blog.jeedom.com/6107-parlons-zigbee/

Explains why Jeedom recommends their zigpy based implementation over alternative solutions for Zigbee and why all are complicated.

**Comment by Hedda** [comment]
@zoic21 Please give another consideration to publishing the full Jeedom Zigbee Plugin on a public Git repo at https://github.com/jeedom ?

Not having that implementation published publicly is hindering other zigpy based projects from learning. Ex. "Zigbee Plugin for Domoticz":

See https://github.com/zigpy/zigpy/discussions/865 (and https://github.com/zigbeefordomoticz/Domoticz-Zigbee/tree/beta6 + https://github.com/zigbeefordomoticz/wiki/tree/zigpy)

Having code for the plugin in a Git repo would allow all developers and projects depending on zigpy to re-use and collaborate much easier.

Also, in the spirit of open sourcve I think that should put the Zigbee Plugin for Jeedom in a public GitHub repository under https://github.com/jeedom as have done with Jeedom itself and other plugins for Jeedom, so that way developers from other projects like can take advantage by learning from that zigpy based implementation as well.

Again, today Jeedom developers do share the source code of the zigpy dependent Zigbee plugin for Jeedon directly via a Git repository. Instead just to those who bought their Zigbee plugin via the Jeedom Market, which while technically allowed according to GPL license used it is not in the spirit of [FOSS (Free and Open Source Software)](https://en.wikipedia.org/wiki/Free_and_open-source_software). 

Since other developers can not simply download the Jeedom plugin from a Git repository they have to jump through some hops to get the source code, as they instead need to first install Jeedom and register account for their Jeedom Market, then buy and install the Zigbee Plugin from their Jeedom Market, after that the plugin can be installed on your computer. As the plugin is coded in Python it is also source code and as such it is now available to you on your local computer from the installation directory:

PS: I do not personally agree with their practice of having a partially closed system which do not make code directly available via Git to other developers as I believe in the spirit of [FOSS (Free and Open Source Software)](https://en.wikipedia.org/wiki/Free_and_open-source_software) when depending on other projects code.

**Reply by zoic21** [reply]
Hello I understand and I agree with you but is not on my hand... This is decide by jeedom sas and I'am not and employee or decider of jeedom sas. I can do nothing sorry.

**Reply by pipiche38** [reply]
@hedda, this seems to be a best practice from Domadoo and Jeedom. Very sad and very against Open Source practices

**Reply by Hedda** [reply]
> Hello I understand and I agree with you but is not on my hand... This is decide by jeedom sas and I'am not and employee or decider of jeedom sas. I can do nothing sorry.

@zoic21 Do you personally have contacts working for Jeedom SAS and if so could maybe point them to this thread for discussion?

> @Hedda, this seems to be a best practice from Domadoo and Jeedom. Very sad and very against Open Source practices

Sad yes, but to clarify, while it is legal to depend on "open source" licenses and only give code to those who buy the software based on it, that practice goes against some of the core FOSS/FLOSS philosophies, especially that of Free Software Foundation and projects that use its [GNU General Public License v3 license](https://github.com/zigpy/zigpy/blob/dev/LICENSE) and "[The Open Source Definition](https://en.wikipedia.org/wiki/The_Open_Source_Definition)" by the [Open Source Initiative (OSI)](https://en.wikipedia.org/wiki/Open_Source_Initiative), as while a software application/plugin can indeed be pay to use the application they still normally freely distribute the source code of that software so that to other developers using those same dependencies can collaborate. In my humble option, the whole point of using open source is to at least actively give back by also redistributing source code back to the upstream projects so also others can re-use it too:

https://www.gnu.org/philosophy/philosophy.html

https://en.wikipedia.org/wiki/The_Open_Source_Definition

https://en.wikipedia.org/wiki/Free_and_open-source_software

Personally, I see no problems with open source libraries being used in commercial applications/plugins for sale as long as a license the code uses allow it and is followed, however, I will still strongly argue that the spirit of the GPL licenses is also to not restrict or hide access to that source code in any way. So please try to convince Jeedom developers to publish it in an easily accessible manner so developers of other projects can get the code, regardless if they are paying customers or not.
 
PS: As zigpy is GPLv3 there is technically nothing legally preventing me from buying their Zigbee plugin and then redistributing it on GitHub, as an application/plugin depending on a GPLv3 library like zigpy to work needs to have a compatible license to allow that.

**Reply by Hedda** [reply]
> > Hello I understand and I agree with you but is not on my hand... This is decide by jeedom sas and I'am not and employee or decider of jeedom sas. I can do nothing sorry.
> 
> @zoic21 Do you personally have contacts working for Jeedom SAS and if so could maybe point them to this thread for discussion?
> 
> > @Hedda, this seems to be a best practice from Domadoo and Jeedom. Very sad and very against Open Source practices
> 
> Sad yes, but to clarify, while it is legal to depend on "open source" licenses and only give code to those who buy the software based on it, that practice goes against some of the core FOSS/FLOSS philosophies, especially that of Free Software Foundation and projects that use its [GNU General Public License v3 license](https://github.com/zigpy/zigpy/blob/dev/LICENSE) and "[The Open Source Definition](https://en.wikipedia.org/wiki/The_Open_Source_Definition)" by the [Open Source Initiative (OSI)](https://en.wikipedia.org/wiki/Open_Source_Initiative), as while a software application/plugin can indeed be pay to use the application they still normally freely distribute the source code of that software so that to other developers using those same dependencies can collaborate. In my humble option, the whole point of using open source is to at least actively give back by also redistributing source code back to the upstream projects so also others can re-use it too:
> 
> https://www.gnu.org/philosophy/philosophy.html
> 
> https://en.wikipedia.org/wiki/The_Open_Source_Definition
> 
> https://en.wikipedia.org/wiki/Free_and_open-source_software
> 
> Personally, I see no problems with open source libraries being used in commercial applications/plugins for sale as long as a license the code uses allow it and is followed, however, I will still strongly argue that the spirit of the GPL licenses is also to not restrict or hide access to that source code in any way. So please try to convince Jeedom developers to publish it in an easily accessible manner so developers of other projects can get the code, regardless if they are paying customers or not.

@zoic21 Have you heard any news or updates in regards to Jeedom SAS changing their minds about sharing their plugin on GitHub? 

Also, do you happen to know if the Jeedom SAS will do the same with Z-Wave JS Plugin? It has a much less restrictive license:

https://blog.jeedom.com/6406-plugin-z-wave-js-en-beta-test/

> PS: As zigpy is GPLv3 there is technically nothing legally preventing me from buying their Zigbee plugin and then redistributing it on GitHub, as an application/plugin depending on a GPLv3 library like zigpy to work needs to have a compatible license to allow that.

FYI, I am considering this to prove a point that doing so is legal too and no point in Jeedom SAS not sharing Zigbee plugin on GitHub.

**Comment by zoic21** [comment]
Hello,
Sorry I don't have any information on that. I just known that a lawer check if Jeedom SAS have the right to do that and it's ok there is nothing illegal on it. Specially because after buying plugin you have access to the code of plugin. I have only french link but you can find information here : 
- https://www.gnu.org/licenses/gpl-faq.fr.html#DoesTheGPLAllowDownloadFee
- https://www.gnu.org/licenses/gpl-faq.fr.html#DoesTheGPLRequireAvailabilityToPublic

For zwavejs I don't work on it, but we use only API so I think it's good.

**Reply by Hedda** [reply]
Yes I know the GPLv3 license technically allows them to do so legally, however, doing so is not in the spirit of FOSS/FLOSS philosophies.

The practice is among FOSS/FLOSS project and communities considered bad form, as the practice is somewhat similar to “tivoization”.

https://en.wikipedia.org/wiki/Tivoization

It is not "open source friendly" as it hinders developers of other projects from freely studying the code to learn how the program works.

The practice prevents open collaboration between developers working on different projects that also use zigpy libraries as dependencies.

Know that “Free software and open-source software” can be commercial, but currently Jeedom Zigbee plugin is "nonfree", and here "free" refers to "freedom" as in freedom of speech (or freely available source code) and not free as in free beer that peope do not have to pay for, so Jeedom's Zigbee plugin can still be software that users pay for but the code should still really be published for easy access to non-paying developers.

All open-source communities also judge projects based on if a derivative fits open-source spirit as well as the legal precise words in a license.

The freedom to study how the program works, for any purpose must be practical, not just theoretical (i.e., no tivoization/lockdown, etc.).

Not providing easy access to the Zigbee plugin code for others to study it becomes an empty pretence rather than a practical reality.

Again, doing so means that Jeedom does not meet the the open source definition according to the Open Source Initiative (OSI).

So if following that definition then it could be argued that Jeedom should stop calling itself open-source software.

https://en.wikipedia.org/wiki/The_Open_Source_Definition

https://www.gnu.org/philosophy/free-sw.en.html

#### _Free software can be commercial_

_“Free software” does not mean “noncommercial.” On the contrary, a free program must be available for commercial use, commercial development, and commercial distribution. This policy is of fundamental importance—without this, free software could not achieve its aims._

_We want to invite everyone to use the GNU system, including businesses and their workers. That requires allowing commercial use. We hope that free replacement programs will supplant comparable proprietary programs, but they can't do that if businesses are forbidden to use them. We want commercial products that contain software to include the GNU system, and that would constitute commercial distribution for a price. Commercial development of free software is no longer unusual; such free commercial software is very important. Paid, professional support for free software fills an important need._

_Thus, to exclude commercial use, commercial development or commercial distribution would hobble the free software community and obstruct its path to success. We must conclude that a program licensed with such restrictions does not qualify as free software._

_A free program must offer the four freedoms to any would-be user that obtains a copy of the software, who has complied thus far with the conditions of the free license covering the software in any previous distribution of it. Putting some of the freedoms off limits to some users, or requiring that users pay, in money or in kind, to exercise them, is tantamount to not granting the freedoms in question, and thus renders the program nonfree._

...

#### _The freedom to study the source code and make changes_

_In order for freedoms 1 and 3 (the freedom to make changes and the freedom to publish the changed versions) to be meaningful, you need to have access to the source code of the program. Therefore, accessibility of source code is a necessary condition for free software. Obfuscated “source code” is not real source code and does not count as source code._

_Source code is defined as the preferred form of the program for making changes in. Thus, whatever form a developer changes to develop the program is the source code of that developer's version._

_Freedom 1 includes the freedom to use your changed version in place of the original. If the program is delivered in a product designed to run someone else's modified versions but refuse to run yours—a practice known as “tivoization” or “lockdown,” or (in its practitioners' perverse terminology) as “secure boot”—freedom 1 becomes an empty pretense rather than a practical reality. These binaries are not free software even if the source code they are compiled from is free._

_One important way to modify a program is by merging in available free subroutines and modules. If the program's license says that you cannot merge in a suitably licensed existing module—for instance, if it requires you to be the copyright holder of any code you add—then the license is too restrictive to qualify as free._

_Whether a change constitutes an improvement is a subjective matter. If your right to modify a program is limited, in substance, to changes that someone else considers an improvement, that program is not free._

_One special case of freedom 1 is to delete the program's code so it returns after doing nothing, or make it invoke some other program. Thus, freedom 1 includes the “freedom to delete the program.”_

#### _How we interpret these criteria_

_Note that criteria such as those stated in this free software definition require careful thought for their interpretation. To decide whether a specific software license qualifies as a free software license, we judge it based on these criteria to determine whether it fits their spirit as well as the precise words. If a license includes unconscionable restrictions, we reject it, even if we did not anticipate the issue in these criteria. Sometimes a license requirement raises an issue that calls for extensive thought, including discussions with a lawyer, before we can decide if the requirement is acceptable. When we reach a conclusion about a new issue, we often update these criteria to make it easier to see why certain licenses do or don't qualify._

**Comment by zoic21** [comment]
Yes I pretty agree with you in perfect world. But since there is employee in Jeedom SAS and this employee want to be pay Jeedom SAS need to earn money. If jeedom sas change plugin-zigbee (or other pay plugin) to public repository users will just pick it up without paying. With that we can close Jeedom SAS, lay off employees and with it Jeedom project.

I understand yours arguments but it's enought, if jeedom project. If the jeedom project has survived until now it is thanks to the company behind otherwise it would have been dead for a very very long time (it would not even have existed I think)

**Comment by Hedda** [comment]
Jeedom SAS now apparently make/sell a smart home hub called "Jeedom Luna" which has a built-in Zigbee radio as well as a Z-Wave radio:

https://blog.jeedom.com/6842-destination-jeedom-luna/

https://www.jeedom.com/fr/box.html

https://www.youtube.com/results?search_query=Jeedom+Luna

https://www.domadoo.fr/fr/box-domotique/6280-jeedom-controleur-domotique-jeedom-luna-z-wave-700-et-zigbee-30.html

https://shop.domo-supply.com/en/smarthome-hub/2058-jeedom-home-automation-controller-jeedom-luna-z-wave-700-and-zigbee-30.html

They apparently also already sold other appliances called "Jeedom Atlas" and "Jeedom Atlas Pro" which had/have optional Zigbee models?

https://www.jeedom.com/en/box.html

https://www.domadoo.fr/fr/box-domotique/5878-jeedom-controleur-domotique-jeedom-atlas-zigbee.html

https://www.domadoo.fr/fr/box-domotique/6241-jeedom-pack-controleur-domotique-jeedom-atlas-zigbee-dongle-usb-z-wave.html

I assume but do not know if that means that it ships with their official Zigbee Plugin for Jeedom which is based on zigpy without extra cost?

https://www.jeedom.com/fr/soft.html#service_pack

https://doc.jeedom.com/en_US/plugins/home%20automation%20protocol/

> Yes I pretty agree with you in perfect world. But since there is employee in Jeedom SAS and this employee want to be pay Jeedom SAS need to earn money. If jeedom sas change plugin-zigbee (or other pay plugin) to public repository users will just pick it up without paying. With that we can close Jeedom SAS, lay off employees and with it Jeedom project.
> 
> I understand yours arguments but it's enought, if jeedom project. If the jeedom project has survived until now it is thanks to the company behind otherwise it would have been dead for a very very long time (it would not even have existed I think)

@zoic21  With Jeedom SAS now selling hardware boxes + having monthly service fees they could maybe reconsider their stand on being more transparent with Zigbee plugin development by perhaps making their Zigbee plugin repositories open to the public like Jeedom core?

**Comment by Hedda** [comment]
FYI, Jeedom will be abandoned their zigpy-plugin in favour of their new MQTT-based "JeeZigbee" plugin which depends on Zigbee2MQTT.

https://market.jeedom.com/index.php?v=d&p=market_display&id=4351

https://doc.jeedom.com/en_US/plugins/automation%20protocol/z2m/

https://doc.jeedom.com/en_US/plugins/automation%20protocol/z2m/beta/

https://doc.jeedom.com/en_US/plugins/automation%20protocol/z2m/beta/changelog

PS: As can be seen, their documentation for their new plugin has simply been copied from the old plugin so it is not currently correct at all.

**Reply by zigbeefordomoticz** [reply]
Jeedom has never been ethical in terms of using zigpy, so I would not complain.
On the other side I keep a comment I made few years ago zigpy is driven by ZHA and HA and try to forget the idea to have be a fully independent library.
